cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/BrokerSetAwareGoal.java

```diff
@@ -0,0 +1,380 @@
+        Similarly, if excluded topics are specified, then for optimizing efficiently, then we select only included topics' replicas.
+        The reason for picking the below selection functions is to abide by the Optimization Options.
+        Other functions like selecting leaders etc are not relevant for BrokerSetAwareGoal.
+        */
+        new SortedReplicasHelper().maybeAddSelectionFunc(ReplicaSortFunctionFactory.selectImmigrants(),
+                                                         optimizationOptions.onlyMoveImmigrantReplicas())
+                .maybeAddSelectionFunc(ReplicaSortFunctionFactory.selectReplicasBasedOnExcludedTopics(excludedTopics),
+                                       !excludedTopics.isEmpty())
+                .trackSortedReplicasFor(replicaSortName(this, tracksReplicasInReverseOrder, tracksOnlyLeaderReplicas), clusterModel);
+
+        BrokerSetDataStore brokerSetDataStore = _balancingConstraint.getBrokerSetDataStore();
+
+        _brokerSetDataHelper = new BrokerSetDataHelper(clusterModel, brokerSetDataStore);
+    }
+
+    /**
+     * Update goal state.
+     * Sanity check: After completion of balancing / self-healing, confirm that replicas of each topic reside fully in a
+     * broker set.
+     *
+     * @param clusterModel        The state of the cluster.
+     * @param optimizationOptions Options to take into account during optimization.
+     */
+    @Override
+    protected void updateGoalState(ClusterModel clusterModel, OptimizationOptions optimizationOptions)
+            throws OptimizationFailureException {
+        // Sanity check: No self-healing eligible replica should remain at a dead broker/disk.
+        GoalUtils.ensureNoOfflineReplicas(clusterModel, name());
+        // Sanity check: No replica should be moved to a broker, which used to host any replica of the same partition on its broken disk.
+        GoalUtils.ensureReplicasMoveOffBrokersWithBadDisks(clusterModel, name());
+        // Sanity check to confirm that the final distribution is broker set aware.
+        ensureBrokerSetAware(clusterModel, optimizationOptions);
+        if (_provisionResponse.status() != ProvisionStatus.OVER_PROVISIONED) {
+            _provisionResponse = new ProvisionResponse(ProvisionStatus.RIGHT_SIZED);
+        }
+        finish();
+    }
+
+    private void ensureBrokerSetAware(ClusterModel clusterModel, OptimizationOptions optimizationOptions)
+            throws OptimizationFailureException {
+        // Sanity check to confirm that the final distribution is brokerSet aware.
+        Set<String> excludedTopics = optimizationOptions.excludedTopics();
+        Map<String, Set<Broker>> brokersByBrokerSet = _brokerSetDataHelper.getBrokersByBrokerSet();
+
+        // Sanity check to check if brokersByBrokerSet map is not empty
+        if (brokersByBrokerSet == null || brokersByBrokerSet.isEmpty()) {
```

Comment:
Can this ever be `null` or empty? -- i.e. shouldn't we ensure that checks for these cases are done during its construction?

Reply:
updated